### PR TITLE
add configureReproducible( Date lastModifiedDate ) API

### PR DESCRIPTION
**Pull request by hboutemy**
code extracted from POC done in Maven Source Plugin [MSOURCES-120](https://issues.apache.org/jira/browse/MSOURCES-120)
now need to test it from Maven Assembly Plugin to see how the proposed API matches more advanced use cases (other formats than jar, execution bit in Unix file mode, perhaps other aspects)

### CODE REVIEW

**Comment by michael-o** [comment]
I will have a look at this tonight.

File: src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java [hunk 1]
```diff
@@ -1168,4 +1169,31 @@ public void setFilenameComparator( Comparator<String> filenameComparator )
     {
         return filenameComparator;
     }
+
+    @Override
+    public void configureReproducible( int sourceDateEpoch )
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java]: I am not fully conviced by this for several reasons:

* Why introduce `sourceDateEpoch` when `setLastModifiedDate` does the same. The normalization can still be performed.
* You take the option away to provide alternative stable modes by default

I would take another approach:

`configureReproducible()` which checks for a LMD set and aligns to UTC, checks for a file comparator, checks for a file and directory mode set. We actually don't care which exact values have been set as long as they have been set. I am still convinced that the caller provides sane values for reproducibility.

File: src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java [hunk 2]
```diff
@@ -833,4 +835,13 @@ protected String getArchiveType()
         return archiveType;
     }
 
+    @Override
+    protected Date convertSourceDateEpochToDate( int sourceDateEpoch )
+    {
+        // timestamp of zip entries at zip storage level ignores timezone: see https://github.com/Stuk/jszip/issues/369
+        Calendar cal = Calendar.getInstance();
```
[the remark this answers is not in the document]
Comment (hboutemy) [reply src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: "trick the default timezone": I'm not ok with this way of saying things, but it looks it's your way to say what I wrote in comments

File: src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java (hunk 1 above)
Comment (hboutemy) [reply src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java]: because there is a standard for Reproducible Builds that I prefer to use instead of reinventing the wheel: https://reproducible-builds.org/specs/source-date-epoch/

I don't understand what "checks for a LMD set and aligns to UTC" means, nor the 3 following lines: does it just mean that you prefer to replace "int sourceDateEpoch" with a "long lastModifiedDate"?

File: src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (michael-o) [reply src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: But that's what you do. `ZipEntry.setTime()` does use the default timezone. You recalculate the date to have it look like UTC for the `ZipOutputStream`.

File: src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java (hunk 1 above)
Comment (michael-o) [reply src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java]: That's just a write up, not even an RFC. I see no standard here, just a guideline.

The current benefit I see in `configureReproducible()` is just the timezone normalization. The rest can already be achieved with builtin options (strictly speaking). I don't like to introduce another last datetime field which serves the same purpose as `setLastModifiedDate()`, but with less precision.

Let me rephrase my counter approach:

```
validateReproducibility() {
if (lmd == null)
 throw new...
// else maybe align, but this should actually happen when the zip entry is written because if I call `getLMD` I get another value back compared to the one I have written.

if (comparator != null)
  throw new ...

if (forcedFileMode == -1)
  throw new...

if (forcedDirMode == -1)
  throw new...

}
```

Moreover, one just could add  `setTimeZone()` and pass `Etc/UTC` and everything will be properly aligned for your last modified date.

Is that a bit clearer?

File: src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hboutemy) [reply src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: I still don't see what to do with your comment

[the remark this answers is not in the document]
Comment (michael-o) [reply src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: Let's put the sourceEpochDate aside. All you do is to modify the timestamp to look like UTC in the default timezone, don't you? It has nothing to do with the sourceEpochDate. You fix a shortcoming of the ZIP spec. Simply convert to UTC. That is what you need to document.

File: src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java (hunk 1 above)
Comment (hboutemy) [reply src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java]: I don't see why your API "validates" anything: it doesn't seem to have the same objective as mine.
My API is to configure the archiver in a reproducible way, defining an order (that costs CPU, then is a choice that I would not do by default) and fixing Unix mode: you can configure the archiver by hand if you want, or you can also write another API that you prefer to create another API.
Regarding timezone, zip is the only format that does not understand timezones: tar does not have this limitation. That why zip implementation overrides default implementation of convertSourceDateEpochToDate(...)

Comment (michael-o) [reply src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java]: I think I need to change the code to show you what I mean. I think we are talking about the same, but you still don't understand me.

Comment (hboutemy) [reply src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java]: +1

File: src/main/java/org/codehaus/plexus/archiver/Archiver.java [hunk 3]
```diff
@@ -415,11 +415,15 @@ ResourceIterator getResources()
...
     /**
+     * Set filename comparator, used to sort file entries when scanning directories since File.list() does not
+     * guarantee any order.
+     *
      * @since 4.2.0
      */
     void setFilenameComparator( Comparator<String> filenameComparator );
 
     /**
+<<<<<<< reproducible
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/Archiver.java]: What is the purpose of `<<<<<<<`?

File: src/main/java/org/codehaus/plexus/archiver/diags/NoOpArchiver.java [hunk 4]
```diff
@@ -414,4 +414,10 @@ public String getOverrideGroupName()
         return null;
     }
 
+    @Override
+    public void configureReproducible( Date outputTimestamp )
+    {
+
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/diags/NoOpArchiver.java]: Empty line should be removed as in `TrackingArchiver`.

File: src/main/java/org/codehaus/plexus/archiver/Archiver.java [hunk 5]
```diff
@@ -458,4 +462,14 @@ ResourceIterator getResources()
...
+
+    /**
+     * Configure the archiver to get reproducible archives (see <a href="https://reproducible-builds.org/>Reproducible
+     * Builds</a>). This will require reproducible archive entries order, defined entries timestamp and reproducible
+     * entries Unix mode.
+     * 
+     * @param outputTimestamp the date to use for archive entries last modified time
+     * @since 4.2.0
+     */
+    void configureReproducible( Date outputTimestamp );
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/Archiver.java]: Why did you call it `outputTimestamp` if you desribe it with "last modified time"? Why not not used `lastModifiedDate`/`lastModifiedTimestamp` right away?

File: src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java [hunk 6]
```diff
@@ -833,4 +835,24 @@ protected String getArchiveType()
         return archiveType;
     }
 
+    @Override
+    protected Date convertOutputTimestamp( Date outputTimestamp )
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: The name is misleading. You cannot call it conversion if you put in a date and get a date back. `normalizeOutputTimestamp` makes more sense here.

File: src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java [hunk 7]
```diff
@@ -1257,4 +1257,37 @@ public String getOverrideGroupName()
...
+        setFilenameComparator( new Comparator<String>()
+        {
+            @Override
+            public int compare( String s1, String s2 )
+            {
+                return s1.compareTo( s2 );
+            }
+        } );
+
+        // 3. ignore file/directory mode from filesystem, since they may vary based on local user umask
```
Comment (plamentotev) [inline src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java]: This can be quite inconvenient if the archive contains executable scripts (not uncommon for Java applications). I think that at least it should be explicitly stated in the configureReproducible java docs - "reproducible entries Unix mode" is not specific enough. Also I was thinking if it would make sense to keep the executable flag. Having a file mask with the executable flag set to true on Unix system is quite uncommon.

Comment (hboutemy) [reply src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java]: yes, executable bit is the most wanted one.
Not sure this bit is really managed when running on Windows.
I don't know what precisely to do
I'll merge the current state as it is a good initial step: please help by providing a PR to improve

File: src/main/java/org/codehaus/plexus/archiver/Archiver.java (hunk 5 above)
Comment (hboutemy) [reply src/main/java/org/codehaus/plexus/archiver/Archiver.java]: good idea: renaming to lastModifiedDate

File: src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java (hunk 6 above)
Comment (hboutemy) [reply src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: renaming to normalizeLastModifiedDate(...) and adding comments to explain the intent of this method

File: src/main/java/org/codehaus/plexus/archiver/diags/NoOpArchiver.java (hunk 4 above)
Comment (hboutemy) [reply src/main/java/org/codehaus/plexus/archiver/diags/NoOpArchiver.java]: using the convention used in this file: there is an empty line in other methods

File: src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java [hunk 8]
```diff
@@ -833,4 +835,24 @@ protected String getArchiveType()
...
+        return new Date( dosToJavaTime( lastModifiedDate.getTime() ) );
+    }
+
+    /**
+     * Converts DOS time to Java time (number of milliseconds since epoch).
+     *
+     * @see java.util.zip.ZipEntry#setTime
+     * @see java.util.zip.ZipUtils#dosToJavaTime
+     */
+    private static long dosToJavaTime( long dtime )
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: Why not call it `long dosTime`?

Comment (hboutemy) [reply src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: done
